### ISSUE TITLE
GE: Track Usage

### ISSUE THREAD

**Issue by joefoxton** [threefoldtech/home#791]
In order for us to bill customers using a pay-as-you-go model, we need to track usage each day. We need to see how much of their reservations are actually utilised, rather than what was reserved.
- For storage, we need to charge for the peak storage utilisation during the month, which implies daily checks.
- For bandwidth, we need to track total bandwidth utilisation perf month.
- For compute time, we need to track total compute time used for each node.

**Comment by joefoxton** [comment]
Partially covered by #720

**Comment by despiegk** [comment]
thanks to our SDK its very easy for farmers to implement such a model themselves,
until the end of summer we do the billing from the explorer,
after summer the certified capacity will be fully enabled and the farmer will have to decide how to do the billing themselves.

There are multiple possibilities for this

1. use something like https://www.whmcs.com/ and link with 3sdk at backend
2. create wizards & backend logic in 3sdk to deal with billing
3. you hire a company who can do this for you, I think tftech or codescalers might be able to do this

Please note that on TFGrid everything is reservation based, otherwise the smart contract for IT could never work (needs to be deterministic). 

Bandwidth metering is indeed something we need to do and this is planned for summer.

we will try to complete our roadmap to give you more insight how this will look like soon.

**Comment by joefoxton** [comment]
Hey Kristof!

We are indeed building our own portal & workflow with Touch4IT. If you are curious, here are the designs: 
https://www.figma.com/proto/wL5nUmVvvledFnceSZeCtZ/G-IT?node-id=1%3A2&viewport=151%2C285%2C0.029137849807739258&scaling=scale-down <https://www.figma.com/proto/wL5nUmVvvledFnceSZeCtZ/G-IT?node-id=1:2&viewport=151,285,0.029137849807739258&scaling=scale-down>

Thanks for the suggestion of WHMCS. We did a deep evaluation on a paid install and found it to not suit our needs for several reasons, primarily: Web hosting focus, too complex for us & customers, sub-optimal user experience, lots of customisation necessary. Its targeted at DNS/web hosting companies, which is a different set of skills, expertise, and UX tolerance than the markets we are targeting. We instead chose to build our own portal, as we think it is cheaper, a more professional.

In the short term, the back end provisioning of workloads ordered via the portal will be manual. This is our MVP to test the market.

We have assessed that the market will only tolerate Pay-as-you-go. So in order for us to bill this way, we need to be able to track usage (actual usage), rather than reservations (potential usage). We also need this to get a sense of how much we need to undersubscribe a customer to ensure they can grow before hitting the upper limit of a reservation. Hopefully, we will also be able to use this data to know when we need to grow a reservation (which is yet to be implemented). Our business does depend on the ability to grow + extend reservations, or even better, create elastic reservations.

A beautiful replacement for all this would be an inbuilt pay-as-you-go model. But until this exists, we’ll need to access usage data from within a reservation, without compromising the customer’s privacy.

Happy to explain more in person if you like!

Joe

**Comment by joefoxton** [comment]
Keep in mind we are billing in fiat currency, with no connection/trading to TFT. So the customer pays fiat using a pay-as-you-go model, and then from our own tokens, we reserve capacity on their behalf.

**Comment by zaibon** [comment]
@joefoxton You might want to read the new specification regarding the capacity pool and reservation system.

This is not a classic pay as you go model but it introduce concept that you might find useful where a user still reserve a certain amount of capacity, but the consumption of this capacity is not really time based but more actually consumption over time. This mean as long as you don't have any workload using the capacity, the capacity stays available to use. It a bit like a counter.

The full specification document is available at: https://github.com/threefoldtech/tfexplorer/blob/master/specs/modify_IT_contract_over_time.md#extension-of-the-capacity-pool

**Comment by joefoxton** [comment]
Fantastic @zaibon , will read this!

However, before any other solutions to this problem are rolled out, we are hoping for a stop gap solution to let us track actual usage within a reservation. This will give us what we need to manually bill monthly via pay-as-you-go, using Stripe Billing.

So how can we (today) discover usage within a reservation, without compromising customer's right to privacy?

**Comment by zaibon** [comment]
>  we are hoping for a stop gap solution to let us track actual usage within a reservation.

I'm afraid this is something you need to implement yourself on top of the reservation system. I'm not entirely sure what exactly you want to track. Storage seems something easy to track from an application point of view. CPU and memory is a bit harder even though not impossible at all. But from a 0-OS perspective there is no reason to track anything more then what is already done and required to make the reservation model to work.

From what I understand you are trying to build a service on top of the grid. So you need to implement your service requirement yourself, the grid just provides capacity. And since you seems to be the one managing the reservation on behalf of your client, it seems to me that any billing/tracking is something custom for your use case.

**Comment by joefoxton** [comment]
@zaibon We are not yet aligned on the problem or solution. Sounds like we need to clarify a bit more...

The facts about selling S3 right now are:
- The market will only accept Pay-as-you-go via fiat currency. So we are building a dashboard + manual/scripted layer to facilitate this.
- In the short term, any S3 orders that come in through our dashboard will need to be provisioned manually. In the medium term (this year), we will want to do all this automatically via our dashboard and associated python scripts. 
- Seeing as extensible reservations do not yet exist, we need to smartly allocate a sufficient amount of storage for our customers to grow into. Then when reservations become extensible, we will need to carefully manage this to ensure we don't underutilise our capacity, making our business unprofitable.
- We should NEVER have access inside our customer's workloads. This protects our customers right to privacy and data ownership/access. Our only role should be to provision the workload, then share the access key and secret with our customer, while never storing the secret key anywhere. We show the secret once only, then never again. Therefore we will never know actual utilisation of a reservation. We will only know how much we reserved.
- Additionally, we need to map reservations to customers. Anyone can reserve capacity on our grid. We need to map reservations to actual business entities to do billing.

**Given the above facts, we need programatic access to actual utilisation data in order to charge our customers, on top of reservation data.**

In order for us to create a Pay-as-you-go layer, we need to access these metrics:
- S3: how much of a reservation has the customer actually used - checked daily, then we bill the customer on the peak usage that month.
- Bandwidth: how much actual bandwidth was used by customer workloads during the month.
- K8s: how many compute hours were actually used each month.

To our knowledge, we don't currently have access to this data. If we do, please enlighten me on how we get it!

**Comment by joefoxton** [comment]
Hey Guys, is there an update on this? Weynand is saying that there is work going on that will provide an example script for us to see zdb resource utilization... would you mind letting us know when we might see this come through?

**Comment by zaibon** [comment]
> an example script for us to see zdb resource utilization

This is worked on right now.
cc @maxux

**Comment by maxux** [comment]
Here is a visual proof-of-concept on what you can get: https://asciinema.org/a/Ka13oZvQOcUX4HypkeYkgl6GH
This is just by using stats provided by zos and zdb, nothing more :)

**Comment by zaibon** [comment]
Link to the script: https://gist.github.com/maxux/3df6b229902fa1401b3a505d45eec8a9

**Comment by joefoxton** [comment]
Fantastic! Testing this now.

**Comment by joefoxton** [comment]
@maxux @zaibon I've been unable to run the script. Any chance you could add some commenting/documentation to guide me on how to fill it in and run it?

**Comment by maxux** [comment]
Did you updated nodes specifications ?
https://gist.github.com/maxux/3df6b229902fa1401b3a505d45eec8a9#file-minio-stats-py-L5-L29

You have to change host, post, channel and namespace for your minio/zdb storage. You can reach all theses information within the reservation. In addition, when doing the reservation, you have to include a statistics definition (a redis endpoint where to send statistics). Do you need more information on how to do that ?

I can simplify the script by auto-fillin data based on a reservation response, but I need to extract correct information :)
Can you maybe give me your reservation link ?

**Comment by joefoxton** [comment]
Hey Guys, I’m on vacation this week, but will come back to this early next week. Thanks for the response!

Simplifying the script based on a reservation response would be amazing, if you get the time!

Here’s the 1PB reservation I just did for Kapsch. FYI. They’ll be heavily test our S3 over the coming days.
https://explorer.grid.tf/explorer/reservations/10260 <https://explorer.grid.tf/explorer/reservations/10260>
Zdbs: https://explorer.grid.tf/explorer/reservations/10259 <https://explorer.grid.tf/explorer/reservations/10259>

Joe

**Comment by maxux** [comment]
Hello,

It's fine. I'm writing the script to build nodes descriptors based on your reservation.
But you have something missing, on minio reservation, you have to provide the `stats_aggregator:` property to get container statistics pushed somewhere.

**Comment by joefoxton** [comment]
I see. How do I set the stats_aggregator: property?

**Comment by maxux** [comment]
How do you send your reservation ?
The property needs to contains an accessible redis server where to push periodic statistics. It can be in another container running redis. To make things easier for me, in a first step, this redis should be publicly accessible.

**Comment by maxux** [comment]
I updated the gist with a new script: https://gist.github.com/maxux/3df6b229902fa1401b3a505d45eec8a9#file-mksetup-py

Running this script as it will write a file (`/tmp/nodes.json`) which will host configuration generated. If will fetch your both reservation url and extract data from them. The script will produce a warning because your minio doesn't have the stats aggregator set.

It will also set all zdbs found on reservation.

When the script is done, you can run the new version of https://gist.github.com/maxux/3df6b229902fa1401b3a505d45eec8a9#file-minio-stats-py which will now read configuration from file `/tmp/nodes.json`. You will get a warning since no minio will be available. But all zdbs will be monitored correctly.

You have 125 zdbs, the script query each zdb one by one which can take some time, an improved version should query all of them in parallel but this is a proof-of-concept :)

**Comment by joefoxton** [comment]
Hey Max,

Now that Azmy has found a fix for our Kapsch POC issue, I have some time to run these scripts, and share them with our dashboard developers.

Running the mksetup.py code as-is results in the following errors:

./mksetup.py: line 1: from: not found
./mksetup.py: line 2: import: not found
./mksetup.py: line 3: import: not found
./mksetup.py: line 4: import: not found
./mksetup.py: line 5: import: not found
./mksetup.py: line 6: import: not found
./mksetup.py: line 8: class: not found
./mksetup.py: line 9: syntax error: unexpected “("

I added the shebang line: #!/usr/bin/python3, and got this:

zero-os:~/minio-stats# ./mksetup.py 
-ash: ./mksetup.py: not found

**Comment by joefoxton** [comment]
Quick note: I ran these on the minio container. So I sshed into the minio head, and then ran the scripts. Is this the right place to run it?

**Comment by joefoxton** [comment]
Nevermind - I worked out that this needs to be run in a threebot shell. It works. Will now try to use the output to run minio stats.

**Comment by joefoxton** [comment]
Got the scripts to run, but got some errors. I had to modify the scripts slightly to get them to run (fixed a couple of typos). Here's the code that I ran, and the results: 

https://gist.github.com/joefoxton/cb67b0776eb8805ed807a987a36a9338

Sorry for all the noise today.

**Comment by maxux** [comment]
Can you show me modifications you did ?

**Comment by joefoxton** [comment]
joefoxton@Joes-iMac ~ % diff ~/threefold/minio-stats/minio-stats.py ~/Desktop/minio-stats.py
5,32d4
< """
< nodes = {
<     'minio': [
<         {
<             'host': '10.241.0.232',
<             'port': 6379,
<             'channel': 'minio-stats',
<             'client': None,
<             'previous': None,
<         }
<     ],
<     'storage': [
<         {
<             'host': '2a02:2788:54:1096:e459:8eff:fe52:1859',
<             'port': 9900,
<             'namespace': '91-1',
<             'client': None,
<         },
<         {
<             'host': '2a02:2788:54:1096:84c5:fcff:fe3c:5dd4',
<             'port': 9900,
<             'namespace': '91-0',
<             'client': None,
<         }
<     ]
< }
< """
< 
joefoxton@Joes-iMac ~ % diff ~/threefold/minio-stats/mksetup.py ~/Desktop/mksetup.py
0a1,2
> #!/usr/bin/python3
> 
74a77
>     setup.fetch("https://explorer.grid.tf/explorer/reservations/10258")

**Comment by joefoxton** [comment]
Update: @maxux's Storage Utilisation Metrics scripts are done. Now we just need the following:

1. Bandwidth usage per reservation; inbound/outbound
2. Compute hours per reservation
3. Access to identity of reserving party to comply with KYC / AML laws

**Comment by joefoxton** [comment]
Any updates?
